### DISCUSSION TITLE
Logic-driven optional state and proper types

### DISCUSSION THREAD

**Discussion by viveleroi** [colinhacks/zod#3042] [Q&A] [answered]
We use zod to validate form fields, but whether specific fields are **required** is driven by our database and business logic. 

I am not finding enough examples to understand how zod expects this process to happen. 

Assume I have a boolean isRequired which comes from our database. Zod has no `setOptional` so I assume I have to build the schema inside a ternary:

```tsx
const isRequired = true

const foo = isRequired ? z.string() : z.string().optional()
```

This works logically, but breaks the typescript.

```ts
type Model = z.infer<typeof foo> // foo is string | undefined *all the time* even when `isRequired` is true
```

Obviously, the ternary is not sufficient for typescript to know that the schema requires the string.

I'll looking for advice on how to better construct a zod schema so that typescript works. Our real schema is an object with lots of fields, not a single string, but the concept applies to everything.

**Comment by JacobWeisenburger** [comment]
perhaps this will help:
https://github.com/colinhacks/zod/discussions/2099

**Reply by viveleroi** [reply]
I don't believe so, because I don't see a way to use discriminated unions in this context. This isn't conditional based on some key we can discriminate with, it's based purely on a boolean provided by the database.

I essentially want to build my schema, conditionally set every single as required (default) or optiona/nullish, and have the final schema/type match what I expect.

Imagine 5-20 properties in a schema object like this:

```tsx
const schema = z.object({
  foo: const foo = isRequired ? z.string() : z.string().optional()
  // 5-20 more keys, as these represent a form. We use react-hook-form
})

type Model = z.infer<typeof schema>
```

**Comment by mass2527** [comment]
It would be helpful if you created a schema separately and then inferred the type from a different schema. Since TypeScript doesn't know what `isRequired` will be, the Model type in your example should be inferred as the type of both schemas.

If you want to avoid this, you have to manually inform TypeScript. If `isRequired` is true, use one schema; if it is false, use another schema.

```typescript
const schemaWithRequired = z.object({
  foo: z.string(),
  ...
});

const schemaWithoutRequired = z.object({
  foo: z.string().optional(),
  ...
});

const isRequired = Math.random() < 0.5;
if (isRequired) {
  const data = schemaWithRequired.parse(...);
  
  // SchemaWithRequired can be used if you want
  type SchemaWithRequired = z.infer<typeof schemaWithRequired>;
} else {
  const data = schemaWithoutRequired.parse(...);

  // SchemaWithoutRequired can be used if you want
  type SchemaWithoutRequired = z.infer<typeof schemaWithRequired>;
}
```

**Reply by viveleroi** [reply]
Unfortunately this doesn't scale to the object shape I refer to. Say I have 20 keys and they can all be optionally required based on our database, that's impossible to write a different shape for every possible combination. Sounds like this is a no-go but I can't imagine everyone using zod can 100% of the time hard-code requirements like this.

**Reply by JacobWeisenburger** [reply]
can you give an real world example of an object where every one of it's keys might be optional or might be required?

**Reply by viveleroi** [reply]
Any object

```ts
const schema = z.object({
  title: z.string(),
  comment: name: z.string(),
  quantity: z.string()
})
```

Any or all of those may optional because our server/database handles it differently depending on business rules. Some forms are this small, some have 20 fields.

**Reply by JacobWeisenburger** [reply, marked as answer]
It seems like you are trying to do something that I have never needed to do. Perhaps there is a way to simplify the problem. If you would like me to take a deep dive into your issue and try to come up with a solution, it will require a [Sponsored Consulting Session](https://github.com/sponsors/JacobWeisenburger?frequency=one-time). Let me know if that's something you would be interested in.
